rust-lang/cargo
# Migrate tests of `cargo-init` to snapbox

### What does this PR try to resolve?

An attempt of migrating tests of `cargo-init` to [snapbox](https://crates.io/crates/snapbox/).

### How should we test and review this PR?

To review this PR, you may compare the old and the new version one by one. Files not listed in `<test-name>.out` are not asserted, so please help me make sure we don't miss any essential file to check.

Some redundant tests are covered by other tests or renamed. 



<details>
<summary>Expand to see the list of merged/renamed tests.</summary>

- `gitignore_appended_not_replaced` -> merged into `git_ignore_exists_no_conflicting_entries`
- `gitignore_added_newline_in_existing` -> merged into `git_ignore_exists_no_conflicting_entries`
- `gitignore_no_newline_in_new` -> merged into `simple_git`
- `terminating_newline_in_existing_git_ignore` -> merged into `simple_git_ignore_exists`
- `terminating_newline_in_new_git_ignore` -> merged into `simple_git`
- `terminating_newline_in_new_mercurial_ignore` -> merged into `simple_hg`
- `terminating_newline_in_existing_mercurial_ignore` -> merged into `simple_hg_ignore_exists`
- `mercurial_added_newline_in_existing` -> merged into `simple_hg_ignore_exists`
- `mercurial_no_newline_in_new` -> merged into `simple_hg`
- `cargo_lock_gitignored_if_lib1` -> merged into `simple_git`
- `cargo_lock_gitignored_if_lib2` -> renamed to `inferred_lib_with_git`
- `cargo_lock_not_gitignored_if_bin2` -> renamed to `inferred_bin_with_git`
- `cargo_lock_not_gitignored_if_bin1` -> renamed to `explicit_bin_with_git`

</details>



### Additional information

I won't say the process of the migration was pleasant, but overall it results to a higher coverage of output file changes. Here are steps I performed to migrate a test case:

1. Run the old test and observe its output layout. Recommend using a separate worktree to preserve the temporary test output files.
2. Read the test code to arrange its input fixture at `<test-name>.in`.
3. Copy the old output layout or hand-pick the output layout you need. You don't need to fill contents of those files. `snapbox` will do it for you.
4. Run `SNAPSHOTS=overwrite cargo test --test testsuite <your-test-filter>` to assert and generate snapshots.
5. Compare the old layout and the new layout to see if anything missing.

Something observations when dealing with the migration:

* snapbox hasn't yet support unordered assertion.
* snapbox cannot assert inexistence of a file (And probably never?).
* No performance hit so far (measured with hyperfine).
